### PR TITLE
ammo reloads and tick example

### PR DESCRIPTION
**Pull request by dekespo**
The ammo "recharges" with the engine's time tick, once it finishes the ammo, you can see it loads slowly unless you press space bar.

The name FWeapon::tick is a bit confusing as I had to use this method in FCharacter::reload (cannot use FCharacter::tick as it belongs to something else). There are FWeapon::tick, FWeapon::reload (this just maximise the available ammo suddenly) and FCharacter::reload. I would like to have a different names (such as FWeapon::recharge instead of FWeapon::tick and  FCharacter::recharge instead of FCharacter::reload so that reload methods will have the same power). What do you think?

### CODE REVIEW

File: src/FGame.cpp [hunk 1]
```diff
@@ -89,6 +89,8 @@ void FGame::processEvents()
 			
 		if (sf::Keyboard::isKeyPressed(sf::Keyboard::Space))
 			it->fire();
+		else if (!sf::Keyboard::isKeyPressed(sf::Keyboard::Space)) 
+			it->reload(m_engine->getTickCount());
```
Comment (naslundx) [inline src/FGame.cpp]: This is a bit overkill, as it means the rendering engine is responsible for reloading the weapons in the game engine.

File: src/FCharacter.cpp [hunk 2]
```diff
@@ -142,6 +142,11 @@ void FCharacter::fire()
 	m_weapons[m_currentWeapon].fire(/*m_facingLeft ? FDirection::left : FDirection::right*/);
 }
 
+void FCharacter::reload(float delta)
+{
+	m_weapons[m_currentWeapon].tick(delta);
+}
```
Comment (naslundx) [inline src/FCharacter.cpp]: This is a great solution to making it frame-invariant, _but_ a more clean way of doing this is to make `FCharacter::tick()` call `tick()` on all weapons in `m_weapons`. That way all weapons are reloaded, not just the current one - and we don't need an extra method. (Maybe or may be not we only want to load the current weapon, that would create very different gameplay, but let's see later)

File: src/FWeapon.cpp [hunk 3]
```diff
@@ -54,4 +54,4 @@ FWeaponType FWeapon::getType()
 FProjectileType FWeapon::getProjectileType()
 {
 	return m_projectileType;
-}
\ No newline at end of file
+}
```
Comment (naslundx) [inline src/FWeapon.cpp:57]: :+1:

**Comment by naslundx** [comment]
Well done with the git thing.:)

If you have time before me you can edit as I suggested (if it makes any sense) and then commit using `git commit --amend` and then `git push -f origin weapon_ammo_bug`

File: src/FWeapon.cpp (hunk 3 above)
Comment (dekespo) [inline src/FWeapon.cpp:57]: I don't know why this even exists :D

File: src/main.cpp [hunk 4]
```diff
@@ -60,4 +60,4 @@ int main(int argc, char** argv)
 	game.run();
 
 	return 0;
-}
\ No newline at end of file
+}
```
Comment (dekespo) [inline src/main.cpp:63]: How do I convince git that I didn't touch the file (besides not overwriting it) ? :)

**Comment by dekespo** [comment]
Do you mean this way in the code?

File: src/FCharacter.cpp [hunk 5]
```diff
@@ -140,6 +142,12 @@ bool FCharacter::facingLeft()
 void FCharacter::fire()
 {
 	m_weapons[m_currentWeapon].fire(/*m_facingLeft ? FDirection::left : FDirection::right*/);
+	m_load = false;
+}
+
+void FCharacter::loadWeapon()
+{
+	m_load = true;
```
Comment (dekespo) [inline src/FCharacter.cpp]: Still I had to add another method :/ , I just couldn't find a way to make m_load back to false automatically once it stops firing. It uses loadWeapon method in FGame.cpp

File: src/FCharacter.cpp (hunk 2 above)
Comment (dekespo) [inline src/FCharacter.cpp]: I think I have done this OK.

File: src/FGame.cpp (hunk 1 above)
Comment (dekespo) [inline src/FGame.cpp]: Yes, agree. Didn't know the difference between the ticks :)

File: src/main.cpp (hunk 4 above)
Comment (naslundx) [inline src/main.cpp:63]: You shouldn't `git add` it before you commit, and then it wouldn't be included. There's nothing to do about it now, but don't worry, this is how it should be. :)

**Comment by naslundx** [comment]
This is how I meant. But I don't see the need for `m_load` at all (and still FGame is needed to make sure the engine reloads the weapons). Just load the weapons always.

File: src/FCharacter.cpp [hunk 6]
```diff
@@ -166,6 +166,9 @@ void FCharacter::tick(float delta)
 	{
 		updateAI();
 	}
+
+	for(auto &weapon: m_weapons)
+		weapon.tick(delta);
```
Comment (dekespo) [inline src/FCharacter.cpp:172]: This would be enough for loading ammo all the time.

**Comment by dekespo** [comment]
Yeah, I was trying to force not to press space all the time :) . Is this what you want? I mean this time I didn't have to add anything into FGame as it will be loaded inside FWeapon::tick()

**Comment by marcusnaslund** [comment]
There we go! :D

**Pull request merged**